File: windows/tests/AINotebook.Core.Tests/Rag/SourceSummarizerTests.cs
using AINotebook.Core;
using AINotebook.Core.Models;
using AINotebook.Core.Rag;
using AINotebook.Core.Storage;
using AINotebook.Core.Tests.Helpers;
using Xunit;

namespace AINotebook.Core.Tests.Rag;

public class SourceSummarizerTests
{
    // SourceSummarizerTests.testSummarizesPersistsAndReturns
    [Fact]
    public async Task SummarizesPersistsAndReturns()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("NB", "");
        var s = store.CreateSource(nb.Id!.Value, SourceType.Text, "X", null, null);
        store.ReplaceChunks(s.Id!.Value, new[]
        {
            new ChunkDraft("alpha apple", 2, null),
            new ChunkDraft("beta banana", 2, null),
        });
        var chat = new MockChatClient("This is ", "the summary.");
        var summarizer = new SourceSummarizer(store, chat, "m");

        var summary = await summarizer.SummarizeAsync(s.Id!.Value);

        Assert.Equal("This is the summary.", summary);
        Assert.Equal("This is the summary.", store.SourceSummary(s.Id!.Value));
        Assert.Single(chat.CapturedMessages);
        Assert.Equal(ChatRole.User, chat.CapturedMessages[0][0].Role);
    }

    // SourceSummarizerTests.testNoChunksReturnsEmptyWithoutCallingModel
    [Fact]
    public async Task NoChunksReturnsEmptyWithoutCallingModel()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("NB", "");
        var s = store.CreateSource(nb.Id!.Value, SourceType.Text, "X", null, null);
        var chat = new MockChatClient("unused");
        var summarizer = new SourceSummarizer(store, chat, "m");

        var summary = await summarizer.SummarizeAsync(s.Id!.Value);

        Assert.Equal("", summary);
        Assert.Empty(chat.CapturedMessages);
        Assert.Null(store.SourceSummary(s.Id!.Value));
    }
}

File: Tests/AINotebookCoreTests/SourceSummarizerTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class SourceSummarizerTests: XCTestCase {

    final class MockChatClient: ChatStreaming, @unchecked Sendable {
        var capturedMessages: [[ChatTurn]] = []
        let tokens: [String]
        init(tokens: [String]) { self.tokens = tokens }
        func stream(model: String, messages: [ChatTurn]) -> AsyncThrowingStream<String, Error> {
            capturedMessages.append(messages)
            let toks = tokens
            return AsyncThrowingStream { continuation in
                Task {
                    for t in toks { continuation.yield(t) }
                    continuation.finish()
                }
            }
        }
    }

    func testSummarizesPersistsAndReturns() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [
                ChunkDraft(text: "alpha apple", tokenCount: 2),
                ChunkDraft(text: "beta banana", tokenCount: 2)
            ]
        )
        let chat = MockChatClient(tokens: ["This is ", "the summary."])
        let summarizer = SourceSummarizer(store: store, chat: chat, chatModel: "m")

        let summary = try await summarizer.summarize(sourceId: s.id!)
        XCTAssertEqual(summary, "This is the summary.")
        XCTAssertEqual(try store.sourceSummary(id: s.id!), "This is the summary.")
        XCTAssertEqual(chat.capturedMessages.count, 1)
        XCTAssertEqual(chat.capturedMessages[0].first?.role, .user)
    }

    func testNoChunksReturnsEmptyWithoutCallingModel() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        let chat = MockChatClient(tokens: ["unused"])
        let summarizer = SourceSummarizer(store: store, chat: chat, chatModel: "m")

        let summary = try await summarizer.summarize(sourceId: s.id!)
        XCTAssertEqual(summary, "")
        XCTAssertTrue(chat.capturedMessages.isEmpty)
        XCTAssertNil(try store.sourceSummary(id: s.id!))
    }
}
